# Exercise 4.8 Grouping and Aggregating

### 1. [Import Libraries](#library)

### 2. [Import Dataframe](#dataframe)

### 3. [Find the aggregated mean of order number by department id](#mean_order_dept)

### 4. [Create a Loyalty Flag](#loyalty)

### 5. [Group Users by average price of all their orders](#group_price)

### 6. [Create a Frequency Flag](#frequency)

### Import Libraries
<a id='library'></a>

In [1]:
import pandas as pd
import numpy as np
import os

### Import dataframe
<a id='dataframe'></a>

In [2]:
path = r"C:\Users\corri\Documents\New Career\Achievment 4\Instacart Basket Analysis\Data"

In [3]:
ords_prds_merge = pd.read_pickle(os.path.join(path, 'Prepared Data', 'grouped_orders.pkl'))

In [4]:
ords_prds_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  aisle_id  department_id  prices       price_ranges  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   
5         Soda        77              7     9.0  Mid-range product   
6         Soda        77              7     9.0  Mid-range product   
7         Soda        77              7     9.0  Mid-range product   
8         Soda        77              7     9.0  Mid-range product   
9         Soda        77              7     9.0  Mid-range product   

      busiest_day shopping_days      busy_hours  max_order  loyalty_flag  
0  Regularly Busy   Regular Day     Most orders         10  New Customer  
1  Regularly Busy  Slowest Days  Average orders         10  New Customer  
2  Regularly Busy  Slowest Days     Most orders         10  New Customer  
3      Least Busy  Slowest Days  Average orders         10  New Customer  
4      Least Busy  Slowest Days     Most orders         10  New Customer  
5  Regularly Busy   Regular Day  Average orders         10  New Customer  
6  Regularly Busy  Busiest Days     Most orders         10  New Customer  
7  Regularly Busy  Busiest Days     Most orders         10  New Customer  
8  Regularly Busy  Busiest Days     Most orders         10  New Customer  
9      Least Busy  Slowest Days     Most orders         10  New Customer

In [5]:
ords_prds_merge.shape

(32433030, 19)

### Find aggregated mean of order_number by department_id
<a id='mean_order_dept'></a>

In [6]:
ords_prds_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.215751
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170371
12               15.887622
13               16.583304
14               16.759763
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#### Analyzing the entire dataframe gives the results for all of the departments.  17 (household goods) now has a higher average (15.6 instead of 11) and 4 (produce) now has a lower mean (17.8 instead of 18.8).  Clearly including all of the results alters the statistics.

### Loyalty flag - was already created when following the exercises in the lesson with the following code:
<a id='loyalty'></a>

In [ ]:
ords_prds_merge['max_order'] = ords_prds_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [ ]:
ords_prds_merge.loc[ords_prds_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal Customer'

In [ ]:
ords_prds_merge.loc[(ords_prds_merge['max_order'] <= 40) & (ords_prds_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular Customer'

In [ ]:
ords_prds_merge.loc[ords_prds_merge['max_order'] <= 10, 'loyalty_flag'] = 'New Customer'

In [8]:
ords_prds_merge['loyalty_flag'].value_counts(dropna = False)

Regular Customer    15890458
Loyal Customer      10293434
New Customer         6249138
Name: loyalty_flag, dtype: int64

In [9]:
# Check basic stats of product prices for the three loyalty categories
ords_prds_merge.groupby('loyalty_flag').agg({'prices':['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal Customer    10.388854  1.0  99999.0
New Customer      13.294634  1.0  99999.0
Regular Customer  12.496419  1.0  99999.0

### Group users by average price of all their orders
<a id='group_price'></a>

In [10]:
ords_prds_merge['avg_spent'] = ords_prds_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [22]:
ords_prds_merge['avg_spent'].head(30)

0     6.367797
1     6.367797
2     6.367797
3     6.367797
4     6.367797
5     6.367797
6     6.367797
7     6.367797
8     6.367797
9     6.367797
10    3.980556
11    3.980556
12    3.980556
13    3.980556
14    3.980556
15    8.372059
16    8.372059
17    8.372059
18    7.012683
19    8.137793
20    8.137793
21    7.444898
22    7.444898
23    6.408876
24    6.408876
25    6.408876
26    6.408876
27    6.408876
28    6.408876
29    6.408876
Name: avg_spent, dtype: float64

In [13]:
ords_prds_merge['avg_spent'].describe()

count    3.243303e+07
mean     1.198133e+01
std      8.319275e+01
min      1.000000e+00
25%      7.391935e+00
50%      7.829483e+00
75%      8.258696e+00
max      2.500542e+04
Name: avg_spent, dtype: float64

In [14]:
# Create flags based on avg_spent (< 10 = low spender, >= 10 = high spender)
ords_prds_merge.loc[ords_prds_merge['avg_spent'] < 10, 'spenders'] = 'Low Spender'

In [15]:
ords_prds_merge.loc[ords_prds_merge['avg_spent'] >= 10, 'spenders'] = 'High Spender'

In [17]:
ords_prds_merge[['user_id', 'avg_spent', 'spenders']].head(30)

user_id  avg_spent     spenders
0         1   6.367797  Low Spender
1         1   6.367797  Low Spender
2         1   6.367797  Low Spender
3         1   6.367797  Low Spender
4         1   6.367797  Low Spender
5         1   6.367797  Low Spender
6         1   6.367797  Low Spender
7         1   6.367797  Low Spender
8         1   6.367797  Low Spender
9         1   6.367797  Low Spender
10       15   3.980556  Low Spender
11       15   3.980556  Low Spender
12       15   3.980556  Low Spender
13       15   3.980556  Low Spender
14       15   3.980556  Low Spender
15       19   8.372059  Low Spender
16       19   8.372059  Low Spender
17       19   8.372059  Low Spender
18       21   7.012683  Low Spender
19       31   8.137793  Low Spender
20       31   8.137793  Low Spender
21       43   7.444898  Low Spender
22       43   7.444898  Low Spender
23       52   6.408876  Low Spender
24       52   6.408876  Low Spender
25       52   6.408876  Low Spender
26       52   6.408876  Low Spender
27       52   6.408876  Low Spender
28       52   6.408876  Low Spender
29       52   6.408876  Low Spender

In [19]:
ords_prds_merge['spenders'].value_counts(dropna = False)

Low Spender     31797576
High Spender      635454
Name: spenders, dtype: int64

In [20]:
ords_prds_merge.shape

(32433030, 21)

### Create a frequency flag based on days_since_prior_order median
<a id ='frequency'></a>

In [23]:
ords_prds_merge['med_frequency'] = ords_prds_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [24]:
ords_prds_merge['med_frequency'].head(30)

0     20.5
1     20.5
2     20.5
3     20.5
4     20.5
5     20.5
6     20.5
7     20.5
8     20.5
9     20.5
10    10.0
11    10.0
12    10.0
13    10.0
14    10.0
15     6.0
16     6.0
17     6.0
18     7.5
19     6.0
20     6.0
21    10.0
22    10.0
23     8.0
24     8.0
25     8.0
26     8.0
27     8.0
28     8.0
29     8.0
Name: med_frequency, dtype: float64

In [26]:
# Create frequency flag: > 20 = non-frequent customer, >10 and <=20 = regular customer, <=10 = Frequent customer
ords_prds_merge.loc[ords_prds_merge['med_frequency'] > 20, 'frequent'] = 'Non-Frequent Customer'

In [28]:
ords_prds_merge.loc[(ords_prds_merge['med_frequency'] > 10) & (ords_prds_merge['med_frequency'] <=20), 'frequent'] = 'Regular Customer'

In [29]:
ords_prds_merge.loc[ords_prds_merge['med_frequency'] <= 10, 'frequent'] = 'Frequent Customer'

In [30]:
ords_prds_merge[['user_id', 'med_frequency', 'frequent']].head(30)

user_id  med_frequency               frequent
0         1           20.5  Non-Frequent Customer
1         1           20.5  Non-Frequent Customer
2         1           20.5  Non-Frequent Customer
3         1           20.5  Non-Frequent Customer
4         1           20.5  Non-Frequent Customer
5         1           20.5  Non-Frequent Customer
6         1           20.5  Non-Frequent Customer
7         1           20.5  Non-Frequent Customer
8         1           20.5  Non-Frequent Customer
9         1           20.5  Non-Frequent Customer
10       15           10.0      Frequent Customer
11       15           10.0      Frequent Customer
12       15           10.0      Frequent Customer
13       15           10.0      Frequent Customer
14       15           10.0      Frequent Customer
15       19            6.0      Frequent Customer
16       19            6.0      Frequent Customer
17       19            6.0      Frequent Customer
18       21            7.5      Frequent Customer
19       31            6.0      Frequent Customer
20       31            6.0      Frequent Customer
21       43           10.0      Frequent Customer
22       43           10.0      Frequent Customer
23       52            8.0      Frequent Customer
24       52            8.0      Frequent Customer
25       52            8.0      Frequent Customer
26       52            8.0      Frequent Customer
27       52            8.0      Frequent Customer
28       52            8.0      Frequent Customer
29       52            8.0      Frequent Customer

In [31]:
ords_prds_merge['frequent'].value_counts(dropna = False)

Frequent Customer        21576963
Regular Customer          7216644
Non-Frequent Customer     3639423
Name: frequent, dtype: int64

In [32]:
# Export dataframe
ords_prds_merge.to_pickle(os.path.join(path, 'Prepared Data', 'grouped_orders.pkl'))